File: minitools/login-browserui.rb
require 'Qt4'
require 'qtwebkit'
require 'uri'
require 'json'

WL_AUTHSTART = "https://login.live.com/oauth20_authorize.srf"
REDIRECT_URL = "https://login.live.com/oauth20_desktop.srf"
CLIENTID = "00000000480BC46C"
INITIAL_SCOPE = "service::skype.com::MBI_SSL"

$starturl = sprintf("%s?client_id=%s&scope=%s&response_type=token&redirect_uri=%s&state=999&locale=en", # TODO: Set the locale parameter correctly
                    WL_AUTHSTART,
                    URI::encode(CLIENTID),
                    URI::encode(INITIAL_SCOPE),
                    URI::encode(REDIRECT_URL)
                   )

urls = []

def urlencoded_to_hash(str)
    pairs = str.split("&")
    return pairs.map do |i|
        m = /^([^=]*)=(.*)$/.match i
        {m[1].to_sym => URI.decode(m[2])}
    end.reduce({}){|m,v| m.merge! v}
end

Qt::Application.new(ARGV) do
    Qt::WebView.new do
        self.load Qt::Url.new($starturl)
        self.connect(SIGNAL('urlChanged(QUrl)')) do |url|
            urls << url.toString
            #Qt::Application.instance.quit if urls.last.start_with? REDIRECT_URL
        end
        show
    end
    exec
end

if urls.last.start_with? REDIRECT_URL
    parmstr = /^[^\#]*\#(.*)$/.match(urls.last)[1]
    puts JSON.dump({status: "success", token: urlencoded_to_hash(parmstr)})
else
    puts %q{{"status": "failure"}}
end

File: minitools/login-browserui/loginbrowserui.cpp
#include "loginbrowserui.h"
#include <QUrl>
#include <QUrlQuery>
#include <QCoreApplication>
#include <QNetworkAccessManager>
#include <QNetworkCookieJar>

#include "jsonwritingcookiejar.h"

QString LoginBrowserUi::WL_AUTHSTART = "https://login.live.com/oauth20_authorize.srf";
QString LoginBrowserUi::REDIRECT_URL = "https://login.live.com/oauth20_desktop.srf";
QString LoginBrowserUi::CLIENT_ID("00000000480BC46C");
QString LoginBrowserUi::INITIAL_SCOPE("service::skype.com::MBI_SSL");

LoginBrowserUi::LoginBrowserUi(QWidget *parent)
    : QMainWindow(parent)
{
	this->cookies = new JsonWritingCookieJar(this);

	QUrlQuery params;
	params.addQueryItem("client_id",CLIENT_ID);
	params.addQueryItem("scope", INITIAL_SCOPE);
	params.addQueryItem("response_type","token");
	params.addQueryItem("redirect_uri",REDIRECT_URL);
	params.addQueryItem("state","999");
	params.addQueryItem("locale","en");

	QUrl startUrl(WL_AUTHSTART);
	startUrl.setQuery(params);

	this->view = new QWebView(this);

	this->view->page()->networkAccessManager()->setCookieJar(cookies);

	this->setCentralWidget(view);

	connect(view, SIGNAL(urlChanged(QUrl)), this, SLOT(urlChanged(QUrl)));
	view->load(startUrl);
}

void LoginBrowserUi::urlChanged(QUrl url)
{
	this->urls.append(url);
	if(url.url().startsWith(REDIRECT_URL))
	{
		QCoreApplication::exit(0);
	}
}

QJsonArray LoginBrowserUi::cookieJson()
{
	return cookies->toJson();
}

LoginBrowserUi::~LoginBrowserUi()
{

}
